device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubReceiveTask.java

```diff
@@ -46,11 +50,11 @@ public void run()
             {
                 synchronized (this.receiveThreadLock)
                 {
-                    if (!this.transport.hasReceivedMessagesToHandle() && !this.transport.isClosed())
+                    while (!this.transport.hasReceivedMessagesToHandle() && !this.transport.isClosed())
                     {
                         // AMQP and MQTT layers will notify the IoTHubTransport layer once a message arrives, and at
                         // that time, this thread will be notified to handle them.
-                        this.receiveThreadLock.wait();
+                        this.receiveThreadLock.wait(FAILSAFE_WAKEUP_PERIOD_MILLISECONDS);
```

Comment:
Adding this failsafe actually wasn't what solved the deadlock. But it is a small cost for a safeguard against other potential deadlocks

Reply:
For comparison, we used to spawn a new thread every 10 milliseconds when a client had no traffic. Now it will just wake up 2 threads and go back to sleep every 10 seconds. This is significantly more efficient on devices even with this failsafe